### PR TITLE
Make user business logic reusable

### PR DESCRIPTION
**Pull request by jbaez**
Make user business logic reusable by using `user.domain` from other services
than `user.service`

This is a simple approach to be able to reuse business logic in related services. These changes include moving the business logic for create / update user from the `user.service` to a new module `user.domain`. User domain then is used in user.service, oauth.service and auth.service.

Move some user logic from user-repository to user.service. The business logic should not be in the repositories, the repository is outside the domain (is an adapter) and it could be hard to test logic inside there.
 
Fix user.service tests not awaiting promises. When we have tests with expectations on Promises (rejects, resolves) we need to `await` then or we would have leaking state between tests.

Add `jest-when` lib to make testing easier. This is a nice library I found which we can use to not having to `mockImplementation` in a specific order for the test to pass

### CODE REVIEW

**Comment by jrasm91** [comment]
How can I use user domain in the cli?

**Comment by jbaez** [comment]
> How can I use user domain in the cli?

We still need to move the all the "domain" (I think basically everything except the controller) to to be shared between server, cli, microservices and machine-learning. That is outside of the scope of this PR. After we can do a PR that deals with that.

The CLI is the same as a Controller, so CLI should be using the service. I can see currently we have a `reset-admin-password.command.ts` CLI command which is using the "TypeORM repository" directly. Instead of doing that, we would use `UserService.updateUser` for changing the admin password.

Thinking a bit ahead on what would be involved in this specific case, we would need to define a `cli user`, so we can use it for the `AuthUserDto` and then update the business logic and tests so we allow the `cli user` to change the admin password

**Comment by jbaez** [comment]
Another thing we need to do is to move the validation to the service. We can do that in another PR.
I would suggest to keep using the NestJS DTOs validation decorators. We would create a `validator` class that we would use in the services. This `validator` would use NestJS `class-validator`, the same way we currently do in the validation tests. Something like this for example:

```typescript
import { validate } from 'class-validator';

export async function getValidatedDto(params, ClassDto) {
  const dto = plainToInstance(ClassDto, params);
  const errors = await validate(dto);
  if (errors.length > 0) {
    throw ValidationError(errors);
  }
  return dto
}
```

**Review by jrasm91 — COMMENTED** [review]
In general I think this does several good things:
- Fix broken async tests
- Adds jest-when, which seems like a good fit when mocking multiple calls
- Refactors business logic out of the user repository
- Refactors duplicate logic out of auth service, oauth service, and user service

I personally don't like having user repository calls split between user domain and user service. I think it introduces a lot of potential for future bugs or duplicate code. I would be down to push _all_ the user-repo calls into the user domain class though.

In terms of the ports and adapters architecture, to me it looks like:
- AuthController, OAuthController, and UserController are "primary adapters"
- AuthService, OAuthService, and UserService are "primary ports"
- UserDomain is "core logic"
- UserRepository (interface) is a secondary port
- UserRepository (implementation) is a secondary adapter

There is a 1-to-1 mapping from ports to adapters, and shared logic got moved out of those ports and into a shared service. This looks good to me. Also business logic got moved out of the UserRepository implementation and into "core logic" this also seems good to me.

In that model it would make sense for the `reset-admin-password` command to come with a port/adapter + `UserDomain`. We already have a "primary adapter" (the command class). We could optionally add a new "primary port" if the existing ones don't work with it. So we could add a "CommandLineService", which might have the reset admin password method, but just call UserDomain.updateUser internally.

Thoughts?

**Comment by jbaez** [comment]
> I personally don't like having user repository calls split between user domain and user service. I think it introduces a lot of potential for future bugs or duplicate code. I would be down to push _all_ the user-repo calls into the user domain class though.
> 

Yes you are right, initially I only had the create/update in the user.domain, but then it felt weird using the get from the repository directly in OauthService and AuthService, so I added the getters. I forgot to update the user.service to use the getters from user.domain. I'll get that updated

> UserDomain is "core logic"
Actually I think that's a good name, what do you think about renaming `UserDomain` to `UserCore`

> In that model it would make sense for the `reset-admin-password` command to come with a port/adapter + `UserDomain`. We already have a "primary adapter" (the command class). We could optionally add a new "primary port" if the existing ones don't work with it. So we could add a "CommandLineService", which might have the reset admin password method, but just call UserDomain.updateUser internally.
> 
> Thoughts?

Ideally the `ResetAdminPasswordCommand` would be the "primary adapter" using the existing userService (web, cli, and tests use the system in the same way).
If the current service method does not work for the ResetAdminPasswordCommand (I think in this case it does), as you said it would be a good call to create a new service but I would limit that service to the user scope (something like "UserCommandService")

File: server/apps/immich/src/api-v1/user/user.domain.ts
Comment (jbaez) [reply server/apps/immich/src/api-v1/user/user.domain.ts:65]: Yeah agree. I'll get this updated

File: server/apps/immich/src/api-v1/user/dto/create-user.dto.ts
Comment (jbaez) [reply server/apps/immich/src/api-v1/user/dto/create-user.dto.ts:29]: Yes we need some kind of object to send between layers. Completely Agree about not needing the `ApiProperty` decorators. I did a copy paste since I wanted to keep the validation annotations since that would definitely be needed in the next step when we move validation to the domain
I'll get that updated

File: server/apps/immich/src/api-v1/user/user.domain.ts
Comment (jbaez) [reply server/apps/immich/src/api-v1/user/user.domain.ts:44]: This is part of the business logic. Checks if we are creating a admin user (i.e. we are passing a `CreateAdminDto`):
I could use a TypeScript narrowing function if is easier to understand. Something like this:
```typescript
function isCreateUserDto(createUserDto: CreateUserDto | CreateAdminDto | CreateUserOauthDto): createUserDto is CreateUserDto {
    return (createUserDto as CreateAdminDto).isAdmin === true
  }
```
I think is better if we have all the business logic related to how we create a user in the same place, instead of scattered in services that use this core domain file.

File: server/apps/immich/src/api-v1/user/user.domain.ts
Comment (jbaez) [reply server/apps/immich/src/api-v1/user/user.domain.ts:52]: For the reason in the previous comments, to restrict the ways we can create a user (avoid the possibility of trying to create a user without a password and without a oauthId for example)

File: server/apps/immich/src/api-v1/user/user.domain.ts
Comment (jbaez) [reply server/apps/immich/src/api-v1/user/user.domain.ts:38]: This is part of the business logic, it does more that just requiring the email.
We have 3 ways of creating a user:
- normal user - with mandatory username and email
- admin user - with mandatory username, email and `isAdmin` === true
- normal user from oauth - with mandatory email and oauthId

File: server/apps/immich/src/api-v1/user/user.domain.ts
Comment (jrasm91) [reply server/apps/immich/src/api-v1/user/user.domain.ts:44]: The only business logic here is that "the first account created needs to be an admin", it's not enforcing what specific dto was sent. In the future if the OAuthDto sent `isAdmin: true` this code would execute the same way, even though the way we use types here is wrong/misleading. So, I don't think the types used here or the type casting make sense, since they don't actually align with the business logic that we're validating.

File: server/apps/immich/src/api-v1/user/user.domain.ts
Comment (jrasm91) [reply server/apps/immich/src/api-v1/user/user.domain.ts:52]: If that's the actual business logic we're trying to enforce then it should be included in the code here, not implied via the types. I don't see any checks in this code that validate the oauthId or password is set.

**Comment by jrasm91** [comment]
UserCore works for me, and I agree with the comments about the CLI. I think it's just a decision on how we choose to do authentication for the command. Right now the assumption is the only way the method can be invoked is if you run it from the server directly (implying admin access), so as long as that is consistent it seems like it would be ok.

File: server/apps/immich/src/api-v1/user/user.domain.ts
Comment (jrasm91) [reply server/apps/immich/src/api-v1/user/user.domain.ts:38]: I guess at the end of the day it doesn't matter much, but it seems like these types are misleading / misaligned with the actual business logic we're enforcing (and how we validate it).

The business logic we're actually enforcing in code is:
1. a unique email is required
2. a local admin must exist or isAdmin must be true
3. If you pass a password it will get hashed.

The business logic you're saying we want is 3 different ways to create a user. Either we need to should update the code to actually enforce _that_ business logic (remove non-whitelisted properties from the dtos and validate the actual dto in this method, requiring one complete, valid dto), or I think we should change the types to be more flexible and reflect the business logic we're enforcing in the code.





Now that I look at it, I think there might be a bug - the local admin check

Comment (jbaez) [reply server/apps/immich/src/api-v1/user/user.domain.ts]: What you are saying makes sense and I kind of agree with you, so I started to look at adding a type like you suggested like this:

```typescript
export interface UserCreateParams {
    email: string;
    oauthId?: string;
    password?: string;
    firstName?: string;
    lastName?: string;
  }  
```

Then with TDD mindset I said OK lets write a test that makes having something like `{ oauthId: undefined, password: undefined }` fail, and of course is not possible to write such a test with our approach of testing from the hexagon boundary (i.e. the service).

It would force me to write a test for the `user.core` to test that behaviour.
So as long as the user.core (or any core module) is used by the services and the services are tested it should be fine.
We validate the business logic I described above through the DTO classes in each service. If in the future we add a `isAdmin` property to the OAuthDto then that behaviour would need to be tested from the OauthService, and the user.core would need to be updated if needed.

**Comment by jbaez** [comment]
@jrasm91 The pending conversations we have open are all related to the same `createUser` method.
Let me know what you think to the response to the first comment, if you agree we can resolve the others.

The only other option is to unit-test separately the user.core `createUser`, which can also be another way of doing it, but to keep it simple I suggest we should try to keep testing the business logic only from the services?

**Comment by jrasm91** [comment]
I'm fine with leaving testing in the other services.

**Comment by jbaez** [comment]
> Hello @jbaez, is this PR ready for review?

Hey @alextran1502, yes it's ready to review.
What we did here is move the logic and calls to the UserRepository to UserCore. As a summary (I will create some charts and documentation to explain it better so we can suggest and agree on next steps) this is what this addition does:
- Core modules (e.g. UserCore) contain the business logic, take Entities as arguments instead of `userId` and return Entities. Makes it easier to reuse between related services
- Services that use a Core module will use it for everything related to that context, so for example  a service using UserCore will not use `UserRegistry` directly
- Added DTOs that are not used in the API (internal use between services and UserCore) but still have validation annotations (this is for next step of moving validation to the services)
- As always we keep testing from the services, so current tests should not break if we need to refactor a service to extract the logic to a `Core` module. (I would suggest if we need to refactor a service that has no tests, we create the tests tests first). This PR also adds `jest-when` to make it easier to test not needing to define the mock calls in a specific order to make the test pass (makes tests more robust and less coupled to implementation).

I just merged from main and fixed the merge conflicts. All the test are passing. Maybe you can help me test everything should work as expected. Note that I moved the recent addition of `fs.access` to `UserCore.getUserProfileImage` since that was a merge conflict in UserService and now is part of UserCore.

Let me know what you think 😃

File: server/apps/immich/src/api-v1/oauth/oauth.service.spec.ts
Comment (jrasm91) [inline server/apps/immich/src/api-v1/oauth/oauth.service.spec.ts:167]: Why is there a call to `getAdmin` for a user that already exists? This test is about updating a user (not setting isAdmin: true), so I'm not sure why this call is happening here.

File: server/apps/immich/src/api-v1/user/dto/create-user.dto.ts
Comment (jrasm91) [reply server/apps/immich/src/api-v1/user/dto/create-user.dto.ts:42]: Can you change this to have a capital `O` please - for consistency.

File: server/apps/immich/src/api-v1/user/user.service.ts
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user.service.ts:112]: These didn't get added anywhere as far as I can tell, so I think there's a problem here. Basically, `userRepository.update()` is implemented with `repository.save()`, so it actually requires the full entity. Fields that are omitted during a "save" are removed, which is why these all are listed here. If you aren't going include these lines (merge the dto into the user somewhere and pass the user to repositroy.update), you should update the repository update implementation to not use save.

**Comment by jbaez** [comment]
@alextran1502 Are have the e2e just been enabled?
It looks like they might be broken for a while. I updated the logic in the user.e2e (admin must be created first) and updated some TS module mappers for jest (had lots of errors of module not found) but there are still errors in the e2e

**Comment by alextran1502** [comment]
@jbaez No they were broken for a while and I wasn't sure how to fix them

File: server/apps/immich/src/api-v1/oauth/oauth.service.spec.ts
Comment (jbaez) [reply server/apps/immich/src/api-v1/oauth/oauth.service.spec.ts:167]: Actually this test is about creating a user.
The oauth service checks if the user exists, and it gets a null `userRepositoryMock.getByEmail.mockResolvedValue(null);`
Then in attempts to create a user by doing `this.userCore.createUser`, which as part of the business logic checks if an admin already exists `userRepositoryMock.getAdmin.mockResolvedValue(user);`
I guess I was lazy here and didn't create an additional `userAdmin` resource to return here so it's different  than the one returned by the `create` mock (maybe it would be more readable, but still that getAdmin call would still be needed) 😬

File: server/apps/immich/src/api-v1/user/user.service.ts
Comment (jbaez) [reply server/apps/immich/src/api-v1/user/user.service.ts:112]: Ah right, that's a good catch, I didn't see that. Why is it doing a save instead of un update 🤔. I trusted the interface of the userRepository that takes a partial entity, so that is not right if is doing a save.
```typescript
update(id: string, user: Partial<UserEntity>): Promise<UserEntity>;
```

Actually I can see in the history that interface was changed a month ago. Before it was:
```typescript
update(user: UserEntity, updateUserDto: UpdateUserDto): Promise<UserEntity>;
```
which works for updating an entity with `save`

I prefer the new interface though, it makes more sense to me. What do you think about changing the repository implementation to use `update` instead of `save`?

**Comment by jbaez** [comment]
> @jbaez No they were broken for a while and I wasn't sure how to fix them

Ok I'll put it in my list to get those e2e tests fixed. After we get this fixed, I think we should configure GitHub to prevent a merge to main if any check is failing, otherwise is harder to know what broke the tests (and of course we lose a bit of protection against an actual bug) 😃

File: server/apps/immich/src/api-v1/user/dto/create-user.dto.ts
Comment (jbaez) [reply server/apps/immich/src/api-v1/user/dto/create-user.dto.ts:42]: Yes of course, sorry I forgot about this comment. I'll change that tomorrow 👍

File: server/apps/immich/src/api-v1/oauth/oauth.service.spec.ts
Comment (jrasm91) [reply server/apps/immich/src/api-v1/oauth/oauth.service.spec.ts:167]: Oh, you're right. I must have misread the diff. I thought this for the update case.

File: server/apps/immich/src/api-v1/user/user.service.ts
Comment (jrasm91) [reply server/apps/immich/src/api-v1/user/user.service.ts:112]: That's fine, but I don't think cascades or relations are saved that way. Probably not an issue yet. Also update doesn't return a full user, so you'd need to requery it afterwards. I do like the interface of "only pass what you want to change". So either query the user, merge the new properties and return the whole thing. Or, do an update, then query the new user and return it. Or, I guess change the interface to not return a user entity.

This is a thing we should decide and then make consistent across all the repos though.

Comment (jrasm91) [reply server/apps/immich/src/api-v1/user/user.service.ts:112]: Ah I think I misinterpreted what update you were referring to. 

I like the interface for an update not requiring passing the full entity, but only the fields that have changed. 

I'm fine if we want to always require the full entity though and just do a save on it, but I don't think the repository should be the one merging the two objects together. Just pass the already merged object to the repo and all it does is save it.

Comment (jbaez) [reply server/apps/immich/src/api-v1/user/user.service.ts:112]: I was checking the other repositories and the interfaces in the UserRepository for create and update before they were changed.
This was create:
```typescript
create(createUserDto: CreateUserDto): Promise<UserEntity> 
```
And this was update:
```typescript
update(user: UserEntity, updateUserDto: UpdateUserDto): Promise<UserEntity> 
```
Originally the repository was designed to be responsible for doing any work in the Entities (updating or creating). When an entity is returned from the repository it shouldn't be mutated by the services, and the services shouldn't create Entities.

- For create we pass in the createDto and the repository uses it to do an insert in the DB and returns an entity (since we are using ORM it actually creates a new Entity object and uses it to insert it. Then it returns a new created Entity)
- For update we pass in the entity that we got from a get (without mutating) and the UpdateDto. Then the repository updates it in the DB (since we are using ORM it actually merges the DTO into the Entity)

However, looking the signature of `save`, it looks like you don't need to pass the full entity, a partial would be fine, so we don't need to do the merge at all, which means we can keep following the same "original" rules and make repositories less restricted and also have less logic
```typescript
Repository<UserEntity>.save<Partial<UserEntity>>(entity: Partial<UserEntity>, options?: SaveOptions | undefined): Promise<Partial<UserEntity> & UserEntity> 
```
Which means that the current signature and implementation should be fine.
And actually I just remember I did some manual testing updating user details with the current logic and it was working fine

I think we can leave it as it is, and agree if we should use this same kind of signatures for the other repositories to make it consistent. In my opinion, it makes sense. Create and update just need `Partial<Entity>`

File: server/apps/immich/src/api-v1/user/user.core.ts
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user.core.ts:107]: IMO this can be removed, although we don't have to do it in this PR.

File: server/apps/immich/src/api-v1/user/user.core.ts
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user.core.ts:125]: IMO this can also be removed as well and was only here (as far as I can tell), because `isAdmin` wasn't on the AuthUserDto.

File: server/apps/immich/src/api-v1/user/user.core.ts
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user.core.ts:142]: IMO this can also be removed as well and was only here (as far as I can tell), because isAdmin wasn't on the AuthUserDto.

File: server/apps/immich/src/api-v1/user/user.core.ts
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user.core.ts:146]: This is great, which I believe makes the same business logic in the user repository unnecessary now. Did you forget to move it or did you leave it there on purpose?

File: server/apps/immich/src/api-v1/user/user-repository.ts
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user-repository.ts:73]: This is also business logic that should be in the user core. Also, this throws a BadRequestException, but the repository.updateUser call is wrapped in a try / catch so this will be suppressed and turned into an InternalServerError :disappointed:.

File: server/apps/immich/src/api-v1/user/user.core.ts
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user.core.ts:94]: Not sure why we need to check for write access in order to read the profile image.

File: server/apps/immich/src/api-v1/user/user.core.ts
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user.core.ts:152]: This catches the bad request thrown from the method and turns it into an internal error.

**Review by jrasm91 — APPROVED** [review]
Looks fine to me with the one exception:

`repository.save` does work in the fact that is does a partial update, so that's great. But, in my testing it does not reload the entity, which means a partial update (save) means a partial entity is returned.

File: server/apps/immich/src/api-v1/user/user.core.ts
Comment (alextran1502) [reply server/apps/immich/src/api-v1/user/user.core.ts:94]: It will crash the server if the file is not there

File: server/apps/immich/src/api-v1/user/user.core.ts
Comment (jbaez) [reply server/apps/immich/src/api-v1/user/user.core.ts:152]: The try starts in line 148, so this only catches errors thrown by TypeORM repository

Comment (jbaez) [reply server/apps/immich/src/api-v1/user/user.core.ts:152]: Ah I just saw your other comment in user-repository.
Yes you are right, that throw shouldn't be in the user-repository in the first-place.
It was there for a while and also this try catch, so this wasn't introduced in this PR.

Actually this is fine, what is wrong is the business logic in the repository

File: server/apps/immich/src/api-v1/user/user-repository.ts
Comment (jbaez) [reply server/apps/immich/src/api-v1/user/user-repository.ts:73]: Yes you are right (although this was already like this, not introduced in this PR). I'll move it to the user.core including the TODO (not sure if this can happen since there is no test implemented for this case)

File: server/apps/immich/src/api-v1/user/user.core.ts
Comment (jbaez) [reply server/apps/immich/src/api-v1/user/user.core.ts:146]: That was already in the UserService, I just moved it to UserCore.
But yeah I can see there is the same logic in the repository, I'll get that removed. Thanks!

**Comment by jbaez** [comment]
> Looks fine to me with the one exception:
> 
> `repository.save` does work in the fact that is does a partial update, so that's great. But, in my testing it does not reload the entity, which means a partial update (save) means a partial entity is returned.

Thanks for testing that. Cool so based on that test, a partial save and update is basically the same, which means we have to do a repository.get to fetch the updated user (not a big deal). 
I'll do that

**Comment by jrasm91** [comment]
Yup, and I figured I'd just point out those other "business logic" and redundant things, even though I understand they weren't necessarily in scope for this PR, which has turned into "put everything back to where it was supposed to be" in addition to the original purpose of "make it shareable".

**Comment by jbaez** [comment]
I left the fetch of the authUser with the TODO, we can address that in another PR 😃

**Comment by jrasm91** [comment]
Looks good. I would have used reload myself. 

https://github.com/typeorm/typeorm/blob/51b2a63d/src/repository/BaseEntity.ts#L61-L65

Let's get this merged!

**Comment by jbaez** [comment]
> Looks good. I would have used reload myself.
> 
> https://github.com/typeorm/typeorm/blob/51b2a63d/src/repository/BaseEntity.ts#L61-L65
> 
> Let's get this merged!

Cool I'll fix that pending conflict and get it merged. Thanks!

The `reload` is in the BaseEntity, which we are not extending, so I don't think we can use it. In my opinion is better if we don't extend it to not make it possible to do DB stuff outside the repository 😃

**Pull request merged**
